p2p/pex/pex_reactor.go

```diff
@@ -464,45 +464,116 @@ func (r *Reactor) ensurePeers() {
 	newBias := cmtmath.MinInt(out, 8)*10 + 10
 
 	toDial := make(map[p2p.ID]*p2p.NetAddress)
+	reserve := make(map[p2p.ID]*p2p.NetAddress)
 	// Try maxAttempts times to pick numToDial addresses to dial
-	maxAttempts := numToDial * 3
+	maxToDialAttempts := numToDial * 3
 
-	for i := 0; i < maxAttempts && len(toDial) < numToDial; i++ {
-		if !r.IsRunning() || !r.book.IsRunning() {
-			return
-		}
+	// Calculate reserve size as 50% of numToDial with a minimum of 10
+	reserveSize := cmtmath.MaxInt(numToDial/2, 10)
+	maxReserveAttempts := reserveSize * 3
 
-		try := r.book.PickAddress(newBias)
-		if try == nil {
-			continue
+	filter := func(ka *knownAddress) bool {
+		attempts, lastDialedTime := r.dialAttemptsInfo(ka.Addr)
+		if r.TooEarlyToDial(ka.Addr, attempts, lastDialedTime) != nil {
+			return false
+		}
+		if r.MaxAttemptsToDial(ka.Addr, attempts) != nil {
+			r.book.MarkBad(ka.Addr, defaultBanTime)
+			return false
+		}
+		if r.Switch.IsDialingOrExistingAddress(ka.Addr) {
+			return false
+		}
+		if _, selected := toDial[ka.Addr.ID]; selected {
+			return false
+		}
+		if _, selected := reserve[ka.Addr.ID]; selected {
+			return false
 		}
-		if _, selected := toDial[try.ID]; selected {
+		return true
+	}
+
+	for i := 0; i < maxToDialAttempts && len(toDial) < numToDial; i++ {
+		prospectivePeer := r.book.PickAddress(newBias, filter)
+		if prospectivePeer == nil {
 			continue
 		}
-		if r.Switch.IsDialingOrExistingAddress(try) {
+		toDial[prospectivePeer.ID] = prospectivePeer
+	}
+
+	for i := 0; i < maxReserveAttempts && len(reserve) < reserveSize; i++ {
```

Comment:
Again, another 30 calls to the same method....

Reply:
https://github.com/cometbft/cometbft/pull/3371/commits/6cf054c314136ca0088bca05ac266bb85b8ee1cf